angular/src/app/admin/dam-repository/service-definitions/service-definition-form/service-definition-form.component.html

```diff
@@ -0,0 +1,228 @@
+        <button mat-button
+                type="button"
+                class="uppercase"
+                color="warn"
+                [attr.data-se]="'btn-remove-interface-' + interfaceObj.value.name"
+                (click)="removeInterface(i)">
+          Remove Interface
+        </button>
+      </div>
+    </mat-expansion-panel>
+  </mat-accordion>
+
+  <h3>
+    <span>Roles</span>
+    <div>
+      <button mat-button
+              type="button"
+              color="primary"
+              class="uppercase"
+              data-se="btn-add-role"
+              (click)="addRole()">
+        Add Role
+      </button>
+    </div>
+  </h3>
+  <mat-accordion multi displayMode="flat">
+    <mat-expansion-panel *ngFor="let role of roles.controls; let i=index"
+                         formArrayName="roles"
+                         [attr.data-se]="'role-'+i">
+      <mat-expansion-panel-header collapsedHeight="5rem"
+                                  expandedHeight="5rem">
+        <ng-container *ngIf="role">
+          <mat-panel-title>
+            <mat-icon class="icon icon-identity"></mat-icon>
+            <span>{{role.value.name || 'Role' }}</span>
+          </mat-panel-title>
+          <div class="invalid-warning" *ngIf="role.invalid">
+            <mat-icon class="icon icon-warning"
+                      matTooltip="Role is invalid">
+            </mat-icon>
+          </div>
+        </ng-container>
+      </mat-expansion-panel-header>
+
+      <div [formGroupName]="i">
+        <div class="row">
+          <mat-form-field class="col">
+            <input placeholder="Name"
+                   required
+                   matInput
+                   formControlName="name"
+                   [attr.data-se]="'inp-role-name-' + i"/>
+            <mat-error *ngIf="form.get('roles').controls[i].get('name').invalid">
+              Name must be 3-46 characters and alphanumeric
+            </mat-error>
+          </mat-form-field>
+          <mat-form-field class="col"
+                          formGroupName="ui">
+            <input matInput
+                   required
+                   placeholder="Display Name"
+                   formControlName="label"
+                   [attr.data-se]="'inp-role-label-' + i"/>
+          </mat-form-field>
+        </div>
+        <mat-form-field formGroupName="ui">
+          <textarea matInput
+                    required
+                    placeholder="Description"
+                    formControlName="description"
+                    [attr.data-se]="'inp-role-description-' + i">
+          </textarea>
+          <mat-error *ngIf="form.get('roles').controls[i].get('ui.description').invalid">
+            Description must be up to 255 characters
+          </mat-error>
+        </mat-form-field>
+        <div class="row">
+          <div class="col">
+            <tag-input formControlName="targetRoles"
+                       secondaryPlaceholder="{{roleSecondaryPlaceholder}}"
```

Comment:
If you want to bind the field value you need to use bracket notation which accepts object, while this one accepts strings.

This is better approach:
```
[secondaryPlaceholder]="isRequired('targetRole') ? 'Add Target Role *' : 'Add Target Role'"
```
Where `isRequired` might look like this:
```
  isRequired(target: string): boolean {
    return _get(this.requirements, target, false);
  }
```

Which allows you to simplify the ugly part in `updateRoleValidations`.